### PR TITLE
#7329 invoker script for convert jar

### PR DESCRIPTION
**Pull request by original-brownbear**
For #7329 and #7326

Offline converter:
* CLI arguments adjusted
* Bash script invocation added
* Tests dried up and adjusted for the new invocation style
* Added lowercase filter for #7326

### CODE REVIEW

File: bin/ingest-convert.sh [hunk 1]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
+
+java -cp "$(cd `dirname $0`/..; pwd)"'/tools/ingest-converter/build/libs/ingest-converter.jar:*' \
```
Comment (suyograo) [inline bin/ingest-convert.sh]: We don't bundle `tools` in our release package. Can you add this dir to https://github.com/elastic/logstash/blob/master/rakelib/artifacts.rake#L8? I would add `tools/ingest-converter/build/libs/*.jar`. We'll then have access to this location.

Also, I tried running

```sh
gradle clean assemble
bin/ingest-convert.sh 
``` 
and it threw:

```
Exception in thread "main" java.lang.NoClassDefFoundError: joptsimple/OptionSpec
	at org.logstash.ingest.Pipeline.main(Pipeline.java:15)
Caused by: java.lang.ClassNotFoundException: joptsimple.OptionSpec
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 1 more
```

File: tools/ingest-converter/src/main/java/org/logstash/ingest/JsUtil.java [hunk 2]
```diff
@@ -43,24 +50,50 @@ public static ScriptEngine engine() {
...
+     */
+    private static String input(final URI uri) throws IOException {
+        if ("file".equals(uri.getScheme())) {
+            return new String(
+                Files.readAllBytes(Paths.get(uri)), StandardCharsets.UTF_8
+            );
+        } else if ("http".equals(uri.getScheme()) || "https".equals(uri.getScheme())) {
+            throw new IllegalArgumentException("http not implemented yet");
+        } else {
+            throw new IllegalArgumentException("--input must be of schema file:// or http(s)://");
```
Comment (suyograo) [inline tools/ingest-converter/src/main/java/org/logstash/ingest/JsUtil.java]: Lets not suggest `http(s)` until we add it :). We can change to `--input must be of schema file://`

**Comment by original-brownbear** [comment]
@suyograo uff yea let me fix the packaging. It works if you run `./gradlew shadowJar`, let me fix this to just be normal :)

**Comment by original-brownbear** [comment]
@suyograo all addressed I think :)

**Comment by suyograo** [comment]
@original-brownbear couple of issues in my testing:

1. no usage output which is not helpful for users.

```sh
suyog at Suyogs-MacBook-Pro-2 in ~/ws/elastic/logstash on pr/7335 [$]
$ bin/ingest-convert.sh
Exception in thread "main" java.lang.NullPointerException
	at org.logstash.ingest.JsUtil.convert(JsUtil.java:69)
	at org.logstash.ingest.Pipeline.main(Pipeline.java:15)
```

We should print usage when no arguments or incomplete arguments are printed. I think JOpt supports printing help. 

2. Mark `input` and `output` as required params and provide feedback if not specified.
3. Output file should be created if it doesn't exist. If it exists, we should wipe it.
4. I created an output file, but couldn't get it to accept it. Complained that file was not found, but it does exist:

```sh
suyog at Suyogs-MacBook-Pro-2 in ~/ws/elastic/logstash on pr/7335 [$]
$ ls -lrt /tmp/ingest/pipeline_out.conf
-rw-r--r--  1 suyog  wheel  0 Jun  6 23:24 /tmp/ingest/pipeline_out.conf

suyog at Suyogs-MacBook-Pro-2 in ~/ws/elastic/logstash on pr/7335 [$]
$ ls -lrt /tmp/ingest/pipeline1.conf
-rw-r--r--  1 suyog  wheel  727 Jun  5 23:52 /tmp/ingest/pipeline1.conf
```

```sh
suyog at Suyogs-MacBook-Pro-2 in ~/ws/elastic/logstash on pr/7335 [$]
$ bin/ingest-convert.sh --input file:///tmp/ingest/pipeline1.json --output file:///tmp/ingest/pipeline_out.conf
Exception in thread "main" java.lang.IllegalStateException: java.nio.file.NoSuchFileException: file:/tmp/ingest/pipeline_out.conf
	at org.logstash.ingest.JsUtil.convert(JsUtil.java:75)
	at org.logstash.ingest.Pipeline.main(Pipeline.java:15)
Caused by: java.nio.file.NoSuchFileException: file:/tmp/ingest/pipeline_out.conf
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
	at java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:434)
	at java.nio.file.Files.newOutputStream(Files.java:216)
	at java.nio.file.Files.write(Files.java:3292)
	at org.logstash.ingest.JsUtil.convert(JsUtil.java:68)
	... 1 more
```

**Comment by original-brownbear** [comment]
@suyograo Wops, let me mark the params as required ... no idea why I missed that.

I thought we don't specify `--output` as `URI` (that was in your initial example)?

**Comment by suyograo** [comment]
> I thought we don't specify --output as URI (that was in your initial example)?

Argh, that was a mistake on my end in the example. Apologies. Let's make it consistent, and to accept file URI.

**Comment by original-brownbear** [comment]
@suyograo kk, URI or URI and path?

**Comment by suyograo** [comment]
@original-brownbear file URI. `--output file:///tmp/blah`

**Comment by original-brownbear** [comment]
@suyograo alright sec :)

**Comment by suyograo** [comment]
Works great otherwise; tested from a release artifact.

**Comment by original-brownbear** [comment]
happy to hear that, should have the usage output ready in a few minutes too

**Comment by original-brownbear** [comment]
@suyograo fixed I think :)

```
bin/ingest-convert.sh
Option (* = required)  Description                                             
---------------------  -----------                                             
* --input <URI>        Input JSON file location URI. Only supports 'file://' as
                         URI schema.                                           
* --output <URI>       Output Logstash DSL file location URI. Only supports    
                         'file://' as URI schema.                              
Exception in thread "main" joptsimple.MissingRequiredOptionsException: Missing required option(s) [input, output]
        at joptsimple.OptionParser.ensureRequiredOptions(OptionParser.java:426)
        at joptsimple.OptionParser.parse(OptionParser.java:400)
        at org.logstash.ingest.JsUtil.convert(JsUtil.java:72)
        at org.logstash.ingest.Pipeline.main(Pipeline.java:15)
```

**Comment by suyograo** [comment]
@original-brownbear weird, I have your latest commit

```sh
suyog at Suyogs-MacBook-Pro-2 in ~/ws/elastic/logstash on pr/7335 [$]
$ bin/ingest-convert.sh
Exception in thread "main" java.lang.NullPointerException
	at org.logstash.ingest.JsUtil.convert(JsUtil.java:69)
	at org.logstash.ingest.Pipeline.main(Pipeline.java:15)

suyog at Suyogs-MacBook-Pro-2 in ~/ws/elastic/logstash on pr/7335 [$]
$ git log
commit 7ba426b7fd312a5967574167867e0e26a91c8cd4
Author: Armin <me@obrown.io>
Date:   Wed Jun 7 08:51:53 2017 +0200

     #7329 added usage output
```

**Comment by original-brownbear** [comment]
@suyograo did you run `gradle assemble again`?

**Comment by suyograo** [comment]
ah recompile.

**Comment by suyograo** [comment]
just realized.

**Comment by suyograo** [comment]
super. LGTM

**Comment by original-brownbear** [comment]
@suyograo alrighty merging :)

**Comment by suyograo** [comment]
don't forget to squash

**Comment by original-brownbear** [comment]
@suyograo thanks for the reminder, that was a close one :D

**Pull request closed without merging**